### PR TITLE
Added max_unavailable_percentage for compute nodes

### PR DESCRIPTION
**Pull request by JadeHayes**
After testing changes with node counts of both 20 and 180, the observed difference between using 80% and 50% for node unavailability during rotations was negligible. To prioritize higher availability of nodes during updates, I opted to maintain a 50% setting. However, I am willing to reconsider this setting based on valid reasons.

Note: The docs say [Terraform 0.12 introduces the null value for conditionally omitting a value](https://www.hashicorp.com/blog/terraform-0-12-conditional-operator-improvements) When trying to set `max_unavailable = null` before the `max_unavailable_percentage`, terraform apply updates the null value to `0`, setting `max_unavailable = 0`, the slowest possible rotation. Moving `max_unavailable` after setting `max_unavailable_percentage`, stopped the 0 unavailable rotation behavior. 

_Executing and Timing Multiple Configurations_

_configuration_
```
# nodes.tfvars

additional_node_groups = null
default_node_groups = {
  compute = {
    ami                   =  ami-07492d7cc3295213c # alternating ami-0adf64856582fb168
    instance_types        = ["t3.micro"]
    volume = {
      size = 100
      type = "gp3"
    }

    availability_zone_ids = ["usw2-az1", "usw2-az2"]
    max_per_az = 25 # 190 for 180 nodes 
    min_per_az = 15 # 170 for 180 nodes 
    desired_per_az = 20 # 180 for 180 nodes 
  }
  gpu = {
    availability_zone_ids = ["usw2-az1", "usw2-az2"]
  }
  platform = {
    availability_zone_ids = ["usw2-az1", "usw2-az2"]
  }
}
```

_steps to recreate_
* Using terraform I spun up the infra, cluster & nodes using the `tf.sh` script.
* Updated the max_unavailable/max_unavailable_percentage in the nodes.tf file to desired limited (listed below)
* ran `./tf.sh nodes plan` & ` time ./tf.sh nodes apply` to apply the changes above
* Updated the AMI in the nodes.tfvars script to ensure the compute nodes would rotate.
* Ran `./tf.sh nodes plan` & `time ./tf.sh nodes apply` and recorded the results below

## 20 nodes
**Baseline**
_27.11.2023 @ 13:59_ | _this is what we currently use, rolling one by one_
**time**:  **47:07.02**
```
 # nodes.tf
     max_unavailable = 1
```

**30% max_unavailable_percentage**
_1.12.2023_ |  _compute nodes max_unavailable_percentage set to 30_
**time**: **14:54**
```
 # nodes.tf

resource "aws_eks_node_group" "node_groups" {
  for_each             = local.node_groups_by_name
  ...
  update_config {
    max_unavailable_percentage = 30
    max_unavailable            =  1 # platform node group only
  }
}
```

**Current setting: 50% max_unavailable_percentage**
 _5.12.2023_ |_compute nodes max_unavailable_percentage set to 50_
**time**: ** 9:57.63 total**
```
 # nodes.tf

resource "aws_eks_node_group" "node_groups" {
  for_each             = local.node_groups_by_name
  ...
  update_config {
    max_unavailable_percentage = each.value.node_group.max_unavailable_percentage
    max_unavailable            = each.value.node_group.max_unavailable
  }
}
```

**80% max_unavailable_percentage**
 _1.12.2023_ | _compute nodes max_unavailable_percentage set to 80_
**time**: **8:16.54**
```
 # nodes.tf

resource "aws_eks_node_group" "node_groups" {
  for_each             = local.node_groups_by_name
  ...
  update_config {
    max_unavailable_percentage = 80
    max_unavailable            = 1 # platform node group only
  }
}
```

## 180 nodes

**Baseline**
_29.11.2023 @ 10:55_ | _this is what we currently use, rolling one by one_
**time**: **6:26:20.53**
```
 # nodes.tf

resource "aws_eks_node_group" "node_groups" {
  for_each             = local.node_groups_by_name
  ...
  update_config {
    max_unavailable  =   1
  }
}
```

**30% max_unavailable_percentage**
_1.12.2023_ | _compute nodes max_unavailable_percentage set to 30_
**time**:**14:42.4**
```
 # nodes.tf

resource "aws_eks_node_group" "node_groups" {
  for_each             = local.node_groups_by_name
  ...
  update_config {
    max_unavailable_percentage = 30
    max_unavailable            = 1 # platform node group only
  }
}
```

**50% max_unavailable_percentage**
_1.12.2023_ | _compute nodes max_unavailable_percentage set to 50_
**time**: **10:34.84**

```
 # nodes.tf

resource "aws_eks_node_group" "node_groups" {
  for_each             = local.node_groups_by_name
  ...
  update_config {
    max_unavailable_percentage = 50
    max_unavailable            =  1 # platform node group only
  }
}
```

**80% max_unavailable_percentage**
_1.12.2023_ | _compute nodes max_unavailable_percentage set to 80_
**time**: **12:00.47**
```
 # nodes.tf

resource "aws_eks_node_group" "node_groups" {
  for_each             = local.node_groups_by_name
  ...
  update_config {
    max_unavailable_percentage = 80
    max_unavailable            = 1 # platform node group only
  }
}
```

### CODE REVIEW

File: modules/nodes/variables.tf [hunk 1]
```diff
@@ -247,6 +250,7 @@ variable "default_node_groups" {
 variable "additional_node_groups" {
   description = "Additional EKS managed node groups definition."
   type = map(object({
+    node_group_type       = optional(string, null)
```
Comment (JadeHayes) [inline modules/nodes/variables.tf:253]: Nodes beyond those explicitly defined as `platform` will undergo a rotation process with a 50% unavailability setting. I wasn't sure how we wanted to handle "additional nodes"

File: modules/nodes/variables.tf [hunk 2]
```diff
@@ -247,15 +253,17 @@ variable "default_node_groups" {
...
-    desired_per_az        = number
-    availability_zone_ids = list(string)
-    labels                = map(string)
+    ami                        = optional(string, null)
+    bootstrap_extra_args       = optional(string, "")
+    instance_types             = list(string)
+    spot                       = optional(bool, false)
+    min_per_az                 = number
+    max_per_az                 = number
+    max_unavailable_percentage = optional(number)
```
Comment (steved) [inline modules/nodes/variables.tf:262]: I think we should default this to 50% too.

File: modules/infra/variables.tf [hunk 3]
```diff
@@ -226,15 +232,17 @@ variable "default_node_groups" {
...
-    availability_zone_ids = list(string)
-    labels                = map(string)
+    ami                        = optional(string, null)
+    bootstrap_extra_args       = optional(string, "")
+    instance_types             = list(string)
+    spot                       = optional(bool, false)
+    min_per_az                 = number
+    max_per_az                 = number
+    max_unavailable_percentage = optional(number)
+    max_unavailable            = optional(number)
```
Comment (Secretions) [inline modules/infra/variables.tf:242]: I think we probably meant to set this one?

Comment (JadeHayes) [reply modules/infra/variables.tf:242]: Whoops, yep!  Just updated

**Review by Secretions — APPROVED** [review]

**Pull request merged**